[bug] Unable to set cmake_config_version_compat property from consumer conanfile

### Environment details

* Operating System+version: Arch Linux
* Compiler+version: clang 16.0.6
* Conan version: 2.0.11
* Python version: 3.11.1


### Steps to reproduce

1. `conan new cmake_lib -d name=foo -d version=0.1.0`
2. Apply the following changes:
```diff
---
diff --git a/CMakeLists.txt b/CMakeLists.txt
index f4cdb97..d491176 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,13 +1,11 @@
 cmake_minimum_required(VERSION 3.15)
 project(foo CXX)
 

+find_package(armadillo 9.800.0)
 
 add_library(foo src/foo.cpp)
 target_include_directories(foo PUBLIC include)


+target_link_libraries(foo armadillo::armadillo)
 
 set_target_properties(foo PROPERTIES PUBLIC_HEADER "include/foo.h")
 install(TARGETS foo)
diff --git a/conanfile.py b/conanfile.py
index 1f8e43a..e6310ec 100644
--- a/conanfile.py
+++ b/conanfile.py
@@ -33,8 +33,12 @@ class fooRecipe(ConanFile):
     def layout(self):
         cmake_layout(self)
 
+    def requirements(self):
+        self.requires("armadillo/12.2.0")
+
     def generate(self):
         deps = CMakeDeps(self)
+        deps.set_property("armadillo", "cmake_config_version_compat", "AnyNewerVersion")
         deps.generate()
         tc = CMakeToolchain(self)
         tc.generate()
```
3. `conan build`
4. Observe that `armadillo-config-versions.cmake` doesn't contain the appropriate configuration for the `AnyNewerVersion` policy:

```bash
$ cat build/Release/generators/armadillo-config-version.cmake
set(PACKAGE_VERSION "12.2.0")

if(PACKAGE_VERSION VERSION_LESS PACKAGE_FIND_VERSION)
    set(PACKAGE_VERSION_COMPATIBLE FALSE)
else()
    if("12.2.0" MATCHES "^([0-9]+)\\.")
        set(CVF_VERSION_MAJOR ${CMAKE_MATCH_1})
    else()
        set(CVF_VERSION_MAJOR "12.2.0")
    endif()

    if(PACKAGE_FIND_VERSION_MAJOR STREQUAL CVF_VERSION_MAJOR)
        set(PACKAGE_VERSION_COMPATIBLE TRUE)
    else()
        set(PACKAGE_VERSION_COMPATIBLE FALSE)
    endif()

    if(PACKAGE_FIND_VERSION STREQUAL PACKAGE_VERSION)
        set(PACKAGE_VERSION_EXACT TRUE)
    endif()
endif()


```

### Logs

```
$ conan build .

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=16
os=Linux
[conf]
tools.build:compiler_executables={'c': '/usr/bin/clang', 'cpp': '/usr/bin/clang++'}

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=16
os=Linux
[conf]
tools.build:compiler_executables={'c': '/usr/bin/clang', 'cpp': '/usr/bin/clang++'}


======== Computing dependency graph ========
Graph root
    conanfile.py (foo/0.1.0): /home/user/projects/test/config-version-test/conanfile.py
Requirements
    armadillo/12.2.0#9e60c960e487af50970eda6b169a7a90 - Cache
    openblas/0.3.20#bc54895c0f1f0a58c8ea414a3be55950 - Cache

======== Computing necessary packages ========
Requirements
    armadillo/12.2.0#9e60c960e487af50970eda6b169a7a90:d1f701e1085136c6bf96d01d466b8251a03bfda5#b3bb81d2ff3a59c782040fd60c10a902 - Cache
    openblas/0.3.20#bc54895c0f1f0a58c8ea414a3be55950:cc2c5af420e7ff29920090410acccedc80665287#981031f2dbb0371a39c4bfdb7bbd9f90 - Cache

======== Installing packages ========

======== Installing packages ========
openblas/0.3.20: Already installed! (1 of 2)
openblas/0.3.20: Setting OpenBLAS_HOME environment variable: /home/user/.conan2/p/b/openbac3f4148410b0/p
armadillo/12.2.0: Already installed! (2 of 2)
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'env_info' used in: openblas/0.3.20
WARN: deprecated:     'cpp_info.names' used in: openblas/0.3.20

======== Finalizing install (deploy, generators) ========
conanfile.py (foo/0.1.0): Calling generate()
conanfile.py (foo/0.1.0): Generators folder: /home/user/projects/test/config-version-test/build/Release/generators
conanfile.py (foo/0.1.0): CMakeToolchain generated: conan_toolchain.cmake
conanfile.py (foo/0.1.0): Preset 'conan-release' added to CMakePresets.json. Invoke it manually using 'cmake --preset conan-release' if using CMake>=3.23
conanfile.py (foo/0.1.0): If your CMake version is not compatible with CMakePresets (<3.23) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/home/user/projects/test/config-version-test/build/Release/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
conanfile.py (foo/0.1.0): CMakeToolchain generated: CMakePresets.json
conanfile.py (foo/0.1.0): CMakeToolchain generated: ../../../CMakeUserPresets.json
conanfile.py (foo/0.1.0): Generating aggregated env files
conanfile.py (foo/0.1.0): Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']

======== Calling build() ========
conanfile.py (foo/0.1.0): Calling build()
conanfile.py (foo/0.1.0): Running CMake.configure()
conanfile.py (foo/0.1.0): RUN: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/user/projects/test/config-version-test/build/Release/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/user/projects/test/config-version-test" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/user/projects/test/config-version-test"
-- Using Conan toolchain: /home/user/projects/test/config-version-test/build/Release/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: C++ Standard 20 with extensions OFF
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The CXX compiler identification is Clang 16.0.6
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning at CMakeLists.txt:4 (find_package):
  Could not find a configuration file for package "armadillo" that is
  compatible with requested version "9.800.0".

  The following configuration files were considered but not accepted:

    /home/user/projects/test/config-version-test/build/Release/generators/armadillo-config.cmake, version: 12.2.0



-- Configuring done (0.3s)
CMake Error at CMakeLists.txt:8 (target_link_libraries):
  Target "foo" links to:

    armadillo::armadillo

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



-- Generating done (0.0s)
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_POLICY_DEFAULT_CMP0091


CMake Generate step failed.  Build files cannot be regenerated correctly.

*********************************************************
Recipe 'conanfile.py (foo/0.1.0)' cannot build its binary
It is possible that this recipe is not Conan 2.0 ready
If the recipe comes from ConanCenter, report it at https://github.com/conan-io/conan-center-index/issues
If it is your recipe, check if it is updated to 2.0
*********************************************************

ERROR: conanfile.py (foo/0.1.0): Error in build() method, line 48
        cmake.configure()
        ConanException: Error 1 while executing


```

--- a/conan/tools/cmake/cmakedeps/templates/config_version.py
+++ b/conan/tools/cmake/cmakedeps/templates/config_version.py
@@ -20,7 +20,7 @@ class ConfigVersionTemplate(CMakeDepsFileTemplate):
 
     @property
     def context(self):
-        policy = self.conanfile.cpp_info.get_property("cmake_config_version_compat")
+        policy = self.cmakedeps.get_property("cmake_config_version_compat", self.conanfile)
         if policy is None:
             policy = "SameMajorVersion"
         if policy not in ("AnyNewerVersion", "SameMajorVersion", "SameMinorVersion", "ExactVersion"):

--- a/conan/tools/cmake/cmakedeps/templates/target_configuration.py
+++ b/conan/tools/cmake/cmakedeps/templates/target_configuration.py
@@ -28,7 +28,8 @@ class TargetConfigurationTemplate(CMakeDepsFileTemplate):
                             for components_target_name in components_targets_names]
 
         is_win = self.conanfile.settings.get_safe("os") == "Windows"
-        auto_link = self.conanfile.cpp_info.get_property("cmake_set_interface_link_directories")
+        auto_link = self.cmakedeps.get_property("cmake_set_interface_link_directories",
+                                                self.conanfile)
         return {"pkg_name": self.pkg_name,
                 "root_target_name": self.root_target_name,
                 "config_suffix": self.config_suffix,

--- a/conan/tools/cmake/cmakedeps/templates/target_data.py
+++ b/conan/tools/cmake/cmakedeps/templates/target_data.py
@@ -183,7 +183,8 @@ class ConfigDataTemplate(CMakeDepsFileTemplate):
         global_cppinfo = self.conanfile.cpp_info.aggregated_components()
         pfolder_var_name = "{}_PACKAGE_FOLDER{}".format(self.pkg_name, self.config_suffix)
         return _TargetDataContext(global_cppinfo, pfolder_var_name, self._root_folder,
-                                  self.require, self.cmake_package_type, self.is_host_windows)
+                                  self.require, self.cmake_package_type, self.is_host_windows,
+                                  self.conanfile, self.cmakedeps)
 
     @property
     def _root_folder(self):
@@ -199,7 +200,8 @@ class ConfigDataTemplate(CMakeDepsFileTemplate):
         for comp_name, comp in sorted_comps.items():
             deps_cpp_cmake = _TargetDataContext(comp, pfolder_var_name, self._root_folder,
                                                 self.require, self.cmake_package_type,
-                                                self.is_host_windows)
+                                                self.is_host_windows, self.conanfile, self.cmakedeps,
+                                                comp_name)
 
             public_comp_deps = []
             for required_pkg, required_comp in comp.parsed_requires():
@@ -260,7 +262,7 @@ class ConfigDataTemplate(CMakeDepsFileTemplate):
 class _TargetDataContext(object):
 
     def __init__(self, cpp_info, pfolder_var_name, package_folder, require, library_type,
-                 is_host_windows):
+                 is_host_windows, conanfile, cmakedeps, comp_name=None):
 
         def join_paths(paths):
             """
@@ -346,7 +348,8 @@ class _TargetDataContext(object):
         if require and not require.run:
             self.bin_paths = ""
 
-        build_modules = cpp_info.get_property("cmake_build_modules") or []
+        build_modules = cmakedeps.get_property("cmake_build_modules", conanfile) or []
         self.build_modules_paths = join_paths(build_modules)
         # SONAME flag only makes sense for SHARED libraries
-        self.no_soname = str((cpp_info.get_property("nosoname") if self.library_type == "SHARED" else False) or False).upper()
+        nosoname = cmakedeps.get_property("nosoname", conanfile, comp_name)
+        self.no_soname = str((nosoname if self.library_type == "SHARED" else False) or False).upper()

--- a/conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py
+++ b/conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py
@@ -628,3 +628,37 @@ class TestCMakeVersionConfigCompat:
                 endif()
             endif()""")
         assert expected in dep
+
+
+def test_cmakedeps_set_property_overrides():
+    c = TestClient()
+    app = textwrap.dedent("""\
+        import os
+        from conan import ConanFile
+        from conan.tools.cmake import CMakeDeps
+        class Pkg(ConanFile):
+            settings = "build_type"
+            requires = "dep/0.1", "other/0.1"
+            def generate(self):
+                deps = CMakeDeps(self)
+                # Need the absolute path inside package
+                dep = self.dependencies["dep"].package_folder
+                deps.set_property("dep", "cmake_build_modules", [os.path.join(dep, "my_module1")])
+                deps.set_property("dep", "nosoname", True)
+                deps.set_property("other::mycomp1", "nosoname", True)
+                deps.generate()
+            """)
+
+    pkg_info = {"components": {"mycomp1": {"libs": ["mylib"]}}}
+    c.save({"dep/conanfile.py": GenConanfile("dep", "0.1").with_package_type("shared-library"),
+            "other/conanfile.py": GenConanfile("other", "0.1").with_package_type("shared-library")
+                                                              .with_package_info(pkg_info, {}),
+            "app/conanfile.py": app})
+    c.run("create dep")
+    c.run("create other")
+    c.run("install app")
+    dep = c.load("app/dep-release-data.cmake")
+    assert 'set(dep_BUILD_MODULES_PATHS_RELEASE "${dep_PACKAGE_FOLDER_RELEASE}/my_module1")' in dep
+    assert 'set(dep_NO_SONAME_MODE_RELEASE TRUE)' in dep
+    other = c.load("app/other-release-data.cmake")
+    assert 'set(other_other_mycomp1_NO_SONAME_MODE_RELEASE TRUE)' in other

--- a/conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py
+++ b/conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py
@@ -597,3 +597,34 @@ class TestCMakeVersionConfigCompat:
         c.run("create .")
         c.run("install --requires=dep/0.1 -g CMakeDeps", assert_error=True)
         assert "Unknown cmake_config_version_compat=Unknown in dep/0.1" in c.out
+
+    def test_cmake_version_config_compatibility_consumer(self):
+        c = TestClient()
+        app = textwrap.dedent("""\
+            from conan import ConanFile
+            from conan.tools.cmake import CMakeDeps
+            class Pkg(ConanFile):
+                settings = "build_type"
+                requires = "dep/0.1"
+                def generate(self):
+                    deps = CMakeDeps(self)
+                    deps.set_property("dep", "cmake_config_version_compat", "AnyNewerVersion")
+                    deps.generate()
+                """)
+
+        c.save({"dep/conanfile.py": GenConanfile("dep", "0.1"),
+                "app/conanfile.py": app})
+        c.run("create dep")
+        c.run("install app")
+        dep = c.load("app/dep-config-version.cmake")
+        expected = textwrap.dedent("""\
+            if(PACKAGE_VERSION VERSION_LESS PACKAGE_FIND_VERSION)
+                set(PACKAGE_VERSION_COMPATIBLE FALSE)
+            else()
+                set(PACKAGE_VERSION_COMPATIBLE TRUE)
+
+                if(PACKAGE_FIND_VERSION STREQUAL PACKAGE_VERSION)
+                    set(PACKAGE_VERSION_EXACT TRUE)
+                endif()
+            endif()""")
+        assert expected in dep

